### PR TITLE
[YUNIKORN-2260] Partition limit is NOT equivalent with the root limit

### PR DESCRIPTION
**Pull request by brandboat**
### What is this PR for?
Apply partition limits to root queue if the limits for the root queue are not defined.

### What type of PR is it?
* [x] - Bug Fix

### Todos
N/A

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-2260

### How should this be tested?
covered by unit test

### Screenshots (if appropriate)
N/A

### Questions:
N/A

### CODE REVIEW

File: pkg/common/configs/configvalidator.go
Comment (chia7712) [inline pkg/common/configs/configvalidator.go:582]: if `rootQueue` is NOT root, we don't need this check, right?

Comment (brandboat) [reply pkg/common/configs/configvalidator.go:582]: `rootQueue` will always be the root, as `checkLimitsStructure` will be triggered after `checkQueuesStructure`, and `checkQueuesStructure` will ensure that the root element in queues is the root queue.
https://github.com/apache/yunikorn-core/blob/526a8db4b61c0db48a3fa6a41c9f4d59481862e7/pkg/common/configs/configvalidator.go#L662-L667

File: pkg/common/configs/configvalidator.go
Comment (chia7712) [inline pkg/common/configs/configvalidator.go:577]: this change is a kind of behavior change, so it deserves a docs. Could you please file a PR as follow-up to reminder user?

File: pkg/common/configs/configvalidator.go
Comment (chia7712) [inline pkg/common/configs/configvalidator.go:583]: this solution is simple and it should work I think. However, this method is not idempotent as the `PartitionConfig` changed by this method will get failed if we pass `PartitionConfig` to this method again. It seems to me that is error-prone. In other words, the validator hate `PartitionConfig` has both partition-level limits and root-level limits, but it may produce such case itself.

**Review by chia7712 — COMMENTED** [review]
@brandboat thanks for this patch. two comments left.

File: pkg/common/configs/configvalidator.go
Comment (brandboat) [reply pkg/common/configs/configvalidator.go]: Thanks for the detailed review, indeed we should wipe out partition limits after applied them to root queue, already addressed that in the latest commit.

File: pkg/common/configs/configvalidator.go
Comment (brandboat) [reply pkg/common/configs/configvalidator.go]: I'll raised another pr after this pull-request merged. Thanks !

File: pkg/common/configs/configvalidator.go
Comment (chia7712) [inline pkg/common/configs/configvalidator.go:760]: It seems this change is redundant?

File: pkg/common/configs/configvalidator.go
Comment (chia7712) [inline pkg/common/configs/configvalidator.go:731]: pardon, is this change related to this PR?

File: pkg/common/configs/configvalidator.go
Comment (chia7712) [inline pkg/common/configs/configvalidator.go:614]: Could you add comment to https://github.com/apache/yunikorn-core/blob/master/pkg/common/configs/config.go#L52 to reminder readers that the field could be empty. Also, we should honor the root-level limits.

File: pkg/common/configs/config_test.go
Comment (chia7712) [inline pkg/common/configs/config_test.go:918]: ok, that changes are the side effect.

or could we accept the root-level limits is equal with partition-level limits? that may be a good way, as the caller can call either root-level limits or partition-level limits.

**Review by chia7712 — COMMENTED** [review]
@brandboat a couple of comments. please take a look.

File: pkg/common/configs/configvalidator.go
Comment (brandboat) [reply pkg/common/configs/configvalidator.go:760]: Doing this to avoid lint error `G601: Implicit memory aliasing in for loop.`

File: pkg/common/configs/configvalidator.go
Comment (brandboat) [reply pkg/common/configs/configvalidator.go:731]: Yes, we need the `checkLimits` inside `checkQueues` to check root queue limits (after partition limits is applied to root queue), so here extracted the method out of `checkQueuesStructure`

File: pkg/common/configs/configvalidator.go
Comment (chia7712) [reply pkg/common/configs/configvalidator.go:760]: this lint error is caused by calling reference of `partition`, right? Is this lint error existent before?

File: pkg/common/configs/config_test.go
Comment (brandboat) [reply pkg/common/configs/config_test.go:918]: > or could we accept the root-level limits is equal with partition-level limits? that may be a good way, as the caller can call either root-level limits or partition-level limits.

I've considered this before, and at that time, I thought making configuration changes was a relatively simple approach without the need to modify manager.go. Anyway, this is a completely different approach. I will submit another PR and let the community decide.

File: pkg/common/configs/configvalidator.go
Comment (chia7712) [reply pkg/common/configs/configvalidator.go:731]: ok, you want to the root queue gets all fixes (limits and inserts) before calling `checkQueues`.

for another, you change the behavior or `checkQueuesStructure` without changing test-related code. It seems `checkQueuesStructure` has no tests, right? If so, could you file jira to complete units for each `checkXXX`

File: pkg/common/configs/configvalidator.go
Comment (brandboat) [reply pkg/common/configs/configvalidator.go:760]: Yes it's exist from the beginning, you can try `./tools/golangci-lint run` and search error `G601: Implicit memory aliasing in for loop`.

File: pkg/common/configs/config_test.go
Comment (chia7712) [reply pkg/common/configs/config_test.go:918]: ummm, you are right. please ignore my previous comments :)

File: pkg/common/configs/configvalidator.go
Comment (chia7712) [reply pkg/common/configs/configvalidator.go:614]: Or we should make sure that partition limits are always equal to root limits? We do NOT undefine the partition limits. Instead, we allow both limits if they are identical.

File: pkg/common/configs/configvalidator.go
Comment (brandboat) [reply pkg/common/configs/configvalidator.go:731]: filed https://issues.apache.org/jira/browse/YUNIKORN-2297

File: pkg/common/configs/configvalidator.go
Comment (brandboat) [reply pkg/common/configs/configvalidator.go:614]: > Instead, we allow both limits if they are identical.

Sounds good, I'll adopt your approach. Many thanks !

**Review by manirajv06 — APPROVED** [review]
looks good

File: pkg/common/configs/configvalidator.go
Comment (chia7712) [inline pkg/common/configs/configvalidator.go:603]: `partitionConfig.Queues[0]` can be replaced by `rootQueue`

File: pkg/common/configs/configvalidator.go
Comment (chia7712) [inline pkg/common/configs/configvalidator.go:613]: ditto

File: pkg/common/configs/configvalidator.go
Comment (chia7712) [inline pkg/common/configs/configvalidator.go:604]: the error message can include the name of current top queue.

**Review by chia7712 — COMMENTED** [review]
@brandboat a couple of minor comments. PTAL

File: pkg/common/configs/configvalidator.go
Comment (brandboat) [reply pkg/common/configs/configvalidator.go:613]: Actually this can't be replaced with `rootQueue`, otherwise the partition limits won't change.

Comment (chia7712) [reply pkg/common/configs/configvalidator.go:613]: oh, I get it now.

Comment (brandboat) [reply pkg/common/configs/configvalidator.go]: ping @chia7712 , already addressed comments in the latest comment. Many thanks~

**Review by chia7712 — APPROVED** [review]
LGTM. will merge it after QA gets passed.

**Pull request closed without merging**
